source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -216,7 +216,348 @@ pub struct EnvoyHttpFilter {
 }
 
 impl EnvoyHttpFilter {
```

Comment:
In my tests, I created a trait for this so that i could construct a mock which kept internal state in a `HashMap`. But I ran into a problem with interior mutability. The problem is that the setters don't take `&mut self`, so the implementation is forced to implement interior mutability with something like `RefCell`. But the getters return references to the data, which a `RefCell` borrow does not allow since the lifetime of that reference is tied to the `RefCell` borrow. The only way i was able to get around this is using `UnsafeCell` (i think I could've used `Rc` as well to extend the lifetime?).

But if the setters took a `&mut self` I think this would not be a problem. I'm very much a newbie when it comes to Rust, so it's very possible that this is just me not knowing the right way to do things.

Reply:
> But if the setters took a &mut self I think this would not be a problem. I'm very much a newbie when it comes to Rust, so it's very possible that this is just me not knowing the right way to do things.

There are many ways to handle this. For starters I would just use https://docs.rs/mockall/latest/mockall/. It's very good and will likely do what you need to do without having to manually write any mock code. (https://docs.rs/mockall/latest/mockall/#reference-return-values). If you do need to write mutating state and still return references then yes this is not going to be possible without moving to unsafe.

With all of that said, I think it's reasonable to make the mutating functions accept &mut self as if we are modeling the underlying Envoy before the stream is not itself thread safe. I think it's fine to require higher layer rust code to use a mutex if needed to access it. I would argue it's unsound to not do this unless the underlying code has a mutex around the operations.